### PR TITLE
MAINT: stats: Implement cdf and ppf as ufuncs for the cosine distribution.

### LINKED ISSUE #13417: 10,000 times speedup for generating random numbers from the cosine distribution
I've been trying to simulate an inhomogeneous Poisson point process with an oscillating intensity function and 
scipy.stats.cosine.rvs has been extremely slow (1 minute for 60,000 events). I implemented the [inverse transform sampling method](https://en.wikipedia.org/wiki/Inverse_transform_sampling): to get a random variate x from a distribution f we generate a random q from [0,1) and then find x from cdf(f)(x) = q. Seems to be much faster.

```
import scipy.stats
import numpy as np

def cosine_rvs(n):
    
    x = np.linspace(-np.pi, np.pi, 1000)
    q = (np.pi + x + np.sin(x)) / (2 * np.pi)        
    rv = np.interp(np.random.rand(n), q, x) 
        
    return rv    

# testing
n = 20_000
%time rv_fast = cosine_rvs(n)
%time rv_slow = scipy.stats.cosine.rvs(size=n) 
scipy.stats.ks_2samp(rv_slow, rv_fast) 
```

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
Closes gh-13417

MAINT: stats: Implement cdf and ppf as ufuncs for the cosine distribution.

The ultimate goal of this change is to add an efficient implementation
of the PPF (i.e. inverse CDF) for the cosine distribution in stats.

While testing the new inverse CDF function, I found that the existing
CDF function suffered from loss of precision near x=-pi.  The CDF is

    F(x) = (pi + x + sin(x))/(2*pi)

For example, if x = -3.141592, a direct translation of that function
using numpy.pi and numpy.sin gives:

    >>> x = -3.141592
    >>> (np.pi + x + np.sin(x))/(2*np.pi)
    -1.9483458130046464e-17

The CDF should never be negative, so that result is clearly nonsense.
Changing the order of the operands still gives a bad result:

    >>> (np.pi + (x + np.sin(x)))/(2*np.pi)
    0.0

In the neighborhood of -pi, the function is O((x+pi)**3), and -3.141592 + pi
is approximately 6.5e-07, so the result should be of the order 1e-21.

One of the main changes here is to create a ufunc, scipy.special._ufuncs._cosine_cdf,
that uses a Pade approximation of the CDF near x=-pi.  With this function, the
above value is computed much more accurately:

    >>> from scipy.special._ufuncs import _cosine_cdf
    >>> _cosine_cdf(x)
    7.406016328569848e-21

That result is, in fact, accurate to full 64 bit precision (checked
with mpmath).

The second big change here is to add a ufunc for the inverse of the CDF.
The new function is scipy.special._ufuncs._cosine_invcdf.

There is no closed-form expression for the inverse of the CDF, so we
have to compute it numerically. The idea is simple: get a good first
guess for the inverse value, and then apply Halley's method to refine
the result.

To get a good first guess, we'll use series expansions of the inverse.
Two series expansions will be used: one at the center of the function
(i.e. at p=0.5), and one at each end of the interval 0 <= p <= 1.
(Technically that's three expansions but the expansions at the ends
are related by symmetry.)

For the series at the ends (i.e. near p=0 and p=1), we can use
the series given in the wikipedia page on Kepler's equation
(https://en.wikipedia.org/wiki/Kepler%27s_equation).  The series
that we need is the expression for E when e = 1, given in the section
"Inverse Kepler equation".  This is the series that begins

    E = s + s**3/60 + s**5/1400 + ...,  with s = (6*M)**(1/3)

Translating that to p=0 gives, for p near 0,

    invcdf(p) = -pi + s +  s**3/60 + s**5/1400 + ...,

with s = (12*pi*p)**(1/3).

For the series in the center of the function, we revert the
Taylor series of x + sin(x) at x=0 and then convert the result
to a Pade approximant.  See the comments in _cosine.c
for more details.

The new functions _cosine_cdf and _cosine_invcdf are used in
the methods _cdf, _ppf, _sf and _isf of scipy.stats.cosine.

The efficient implementation of the _ppf method means we can
remove 'cosine' from the list of slow distributions in the file
stats/tests/test_continuous.py.

### CODE REVIEW

**Comment by person142** [comment]
Generally Newton's method is overkill for the inverse cdf; we should be able to do this more efficiently/accurately using minimax approximations of the inverse. But, that's also more difficult unless someone happens to have an implementation of the Remez algorithm handy.

Compare e.g. to how Boost implements the inverse error function:

https://www.boost.org/doc/libs/1_64_0/libs/math/doc/html/math_toolkit/sf_erf/error_inv.html

**Comment by WarrenWeckesser** [comment]
> Compare e.g. to how Boost implements the inverse error function:

From the link:

>    These functions use a rational approximation devised by JM to calculate an initial   approximation to the result that is accurate to ~10-19, then only if that has insufficient accuracy compared to the epsilon for T, do we clean up the result using Halley iteration. 

That's basically what I'm doing here--the Padé approximant is one of many kinds of rational approximations.  I haven't, however, measured precisely how accurate the approximation is.  I use Newton's method instead of Halley iteration.  The initial guess should be good enough that only a few iterations of Newton should get close to machine precision accuracy.  The Newton iteration is pretty trivial, and I didn't even include a test for convergence.  It just does four iterations.

**Comment by WarrenWeckesser** [comment]
>  But, that's also more difficult unless someone happens to have an implementation of the Remez algorithm handy.

It would be useful to have a set of tools for building approximations that would allow you to experiment with an assortment of techniques, e.g. Taylor polynomials, Padé approximants, Remez exchange algorithm, Chebyshev approximations (maybe something like [chebfun](http://www.chebfun.org/)?), etc.  I don't know if anything like that already exists.

**Comment by person142** [comment]
Note that boost only uses Halley for arbitrary precision. For double it’s just the rational approximations.

**Comment by person142** [comment]
Chebyshev/Taylor/Pade you can do pretty well with Mpmath. (Do high precision finite differences for Taylor series; you can find some examples in `special/_precompute`.) I don’t know of a Python solution for Remez, I’ve been meaning to write one for my own special functions library...

**Comment by WarrenWeckesser** [comment]
I'm doing some tweaking of the code, so I added WIP to the title.

**Comment by WarrenWeckesser** [comment]
I experimented a bit more, but nothing I tried was any better than what I have now.  Asymptotic expansions are used to get a good guess for the inverse CDF, and Halley's method is used to polish it off.  In the most recently pushed version, I also significantly improved the CDF near -π; the most important change came from noting that `np.pi` does not equal π.

This implementation is now significantly better than the existing implementation of the cosine distribution, but it does not guarantee that the CDF and inverse CDF are computed to machine precision.  Achieving machine precision would require significantly more work (for me, anyway), and there are higher priority issues than that.  I'm OK with making it *better* without making it *perfect*. 😄

**Comment by WarrenWeckesser** [comment]
The CircleCI failure is https://github.com/scipy/scipy/issues/10261.

**Comment by chrisb83** [comment]
Hi Warren,

this looks really nice. I am not very familiar with the approximation of  functions and I wanted to try out a similar approach for the inverse of the cdf of `stats.semicircular`, as a first step to learn a bit more about the topic.  Do you know a good reference about this topic? And could you share the code that computes the coefficient of inverse of the Taylor polynomial? I did not find this in `mpmath` / `sympy`. Thanks!

**Comment by chrisb83** [comment]
Cross-ref #10106

**Comment by WarrenWeckesser** [comment]
@chrisb83, thanks.  I don't know of any specific references off the top of my head, but I'm sure they're out there.  Josh might have some recommendation (ping @person142).

The hammer I often use for these nails is the Padé approximant.  It isn't *optimal*, but it is simple to compute with mpmath and it can work very well.  Even better would be optimal polynomial or rational approximation (i.e. the Remez exchange algorithm that Josh mentioned up-thread), but I haven't tried to put together a toolkit for computing these.  The function `scipy.signal.remez` designs optimal filters with the exchange algorithm, but I don't know if that code could be used for more general problems.

For lack of a better place, I'll put the Taylor series reversion code in [`mpsci`](https://github.com/WarrenWeckesser/mpsci).  In fact, Josh already created a different implementation in scipy/special/_precompute/utils.py several years ago; see the function `lagrange_reversion`.

**Comment by WarrenWeckesser** [comment]
> For lack of a better place, I'll put the Taylor series reversion code in mpsci.

I added the functions `inverse_pade`, `inverse_taylor` and `revert` to the module  `mpsci.polyapprox`, but Sphinx is kicking my butt so I haven't yet updated the documentation at https://warrenweckesser.github.io/mpsci/.  The functions are defined in the file https://github.com/WarrenWeckesser/mpsci/blob/master/mpsci/polyapprox/_inverse_approximant_tools.py

**Comment by WarrenWeckesser** [comment]
> I haven't yet updated the documentation at https://warrenweckesser.github.io/mpsci/. 

Now updated; here are the functions in the `polyapprox` subpackage: https://warrenweckesser.github.io/mpsci/polyapprox.html

File: scipy/special/_cosine_distr.pxd
Comment (person142) [inline scipy/special/_cosine_distr.pxd]: Instead of documenting here, you could add an entry to `special._precompute` that outputs the desired coefficients.

File: scipy/special/_cosine_distr.pxd
Comment (person142) [inline scipy/special/_cosine_distr.pxd]: It looks like this is evaluating `h * p(h^2)` for a polynomial `p`, in which case you can use Cephes' `polevl.h` or add a real polynomial evaluation method to `special._evalpoly.pxd`. The latter is probably preferable because then we can make use of `sc_fma` when `fma`s are available.

File: scipy/special/_cosine_distr.pxd
Comment (person142) [inline scipy/special/_cosine_distr.pxd]: Note that `special._precompute` has a canned method for reverting series: https://github.com/scipy/scipy/blob/master/scipy/special/_precompute/utils.py#L20

File: scipy/special/add_newdocs.py
Comment (person142) [inline scipy/special/add_newdocs.py]: Feels strange to be arguing against documentation but... since this is an internal function maybe it should just link to the stats methods it's going to be used for?

**Review by person142 — COMMENTED** [review]
Looks good! As detailed in a comment, let's clean up the polynomial evaluations.

**Comment by WarrenWeckesser** [comment]
@person142, thanks for the review.  I probably won't get back to this for a couple weeks (in case you're wondering).

File: scipy/special/_cosine_distr.pxd
Comment (WarrenWeckesser) [reply scipy/special/_cosine_distr.pxd]: I added a short script with this code to the `_precomute` directory.

File: scipy/special/_cosine_distr.pxd
Comment (WarrenWeckesser) [reply scipy/special/_cosine_distr.pxd]: I converted the Cython code to C (it was all just pure floating point calculation, so there was no need for Cython), and in that changed, I updated all the polynomial evaluations to use `polevl`.

File: scipy/special/add_newdocs.py
Comment (WarrenWeckesser) [reply scipy/special/add_newdocs.py]: The functions will be used within the distribution `scipy.stats.cosine`, and these (for better or worse) already have standard docstrings for the `cdf` and `ppf` methods.  Is there something *wrong* with providing docstrings here? (I don't think there is!)

**Comment by WarrenWeckesser** [comment]
I'm reviving this PR.  In an update, I converted the implementation of the core calculations from Cython to plain C.  There was nothing in the Cython code that benefited from Cython--it is all scalar floating point calculations.  For such code, I find C easier to work with than Cython.

I addressed several of the old comments.  Comment if there are still issues that you think should be fixed.

The code still creates private ufuncs using the infrastructure provided in `scipy.special`.  This PR could be a test case for https://github.com/scipy/scipy/pull/10281, if/when that work is restarted.

File: scipy/special/_cosine.c
Comment (mdhaber) [inline scipy/special/_cosine.c:95]: Just to check myself: the idea is that even though (M_PI + 1.2246467991473532e-16) may not be distinguishable from 1.2246467991473532e-16 (at least not with much precision), this stuff is specifically written for when (x + M_PI) is small enough that the addition of 1.2246467991473532e-16 can be resolved and makes a difference.

File: scipy/special/_cosine.c
Comment (mdhaber) [inline scipy/special/_cosine.c:286]: The idea is that you'd prefer users to get NaN (and maybe report a bug) than to potentially get an inaccurate value?

File: scipy/special/_cosine.c
Comment (WarrenWeckesser) [reply scipy/special/_cosine.c:95]: Yes, exactly.

File: scipy/special/_cosine.c
Comment (WarrenWeckesser) [reply scipy/special/_cosine.c]: Yes.  However, I did some more testing, and found that on a grid of 10 million points in the interval [0.0018, 0.42], only one step of Halley's method was needed, so I removed the loop and the check.  Now it just does the single Halley step.

**Comment by WarrenWeckesser** [comment]
I removed the "needs work" label.  I'm not sure if we have a protocol (formal or informal) for who removes it and when it should be removed.  If anyone thinks this PR still needs the label (for changes beyond the usual small tweaks that might come up in new reviews), feel free to reapply the label.

**Comment by WarrenWeckesser** [comment]
The code coverage failure is for files that are not run during the tests.

**Comment by mdhaber** [comment]
I am not going to be able to comment on the math or C code, but I did some accuracy testing using `mpmath` as ground truth, and things look really good.

Here is the _relative_ error of `cosine.cdf` (compared to `mpmath` with `dps=200`) over the support of the distribution:

image

The horizontal axis is sort of a symmetric log-spacing - the details aren't important, but `-15` means `1e-15` from the left bound (-pi) and `15` means `1e-15` away from the right bound (pi). The relative error of `cosine.cdf` in `master` is really bad near the left bound because it settles at a minimum value `-1.94908592e-17` even though the true value gets a lot smaller than that. This PR solves that problem. There isn't a huge difference in speed here; evaluating these 200 points shown takes the PR 192 µs vs 207 µs in master.

Here is the relative error of `cosine.pmf` (compared to `mpmath` using `findroot` with `dps=20`, as that's all the precision that's needed) on (0, 1):

image

Again, this PR is much, much better near the bounds. And of course, the speedup is even more impressive: these 200 points take the PR only 328 µs whereas in master it takes a whopping 1.21 s.

So I'll let others comment on the math/code, but I'll go ahead and indicate approval based on the test and cursory review of the code. Seems safe to me, as there's not much variety in the sorts of inputs this function has to handle, so it seems unlikely that there are hidden bugs that wouldn't have been found by all the number crunching.

**Review by mdhaber — APPROVED** [review]
LGTM based on accuracy and speed. Tests are sufficient

**Comment by WarrenWeckesser** [comment]
Nice plots, @mdhaber, thanks!

**Comment by WarrenWeckesser** [comment]
The `codecov` test reported a failure, but the file is not one that is run during the tests.  Is there a way to disable `codecov` for such files?

**Comment by WarrenWeckesser** [comment]
Also, if/when this is merged, I need to make a small correction in the first (long) commit message.  The message refers to the file `_cosine_distr.pxd`, but in the switch to C, that file was removed.  Instead, the comment should refer to `_cosine.c`.  This can be fixed during a squash of the commits.

**Comment by WarrenWeckesser** [comment]
To fix the merge conflict and to correct the commit message, I squashed the commits and force-pushed.  That also fixed the failure that the code coverage check was reporting.  Given one explicit approval and a few more approving comments, I think this is ready to merge.

**Comment by rlucas7** [comment]
> To fix the merge conflict and to correct the commit message, I squashed the commits and force-pushed. That also fixed the failure that the code coverage check was reporting. Given one explicit approval and a few more approving comments, I think this is ready to merge.

I think so too. I've not verified the constants in the Pade approximant but everything else looks good-one question though: 

This distribution is not supported in boost? 

I didn't see it when I looked and wanted to confirm

**Comment by WarrenWeckesser** [comment]
Thanks @rlucas7.  No, this distribution is not in Boost.  The distributions in Boost 1.75 are listed at  https://www.boost.org/doc/libs/1_75_0/libs/math/doc/html/math_toolkit/dist_ref/dists.html

**Review by rlucas7 — APPROVED** [review]
@WarrenWeckesser thanks for the confirmation on lack of boost support for this distribution. 

I'm marking approved, given @mdhaber has also given approval this seems ready to merge but I'll leave open until next week and merge on Sunday (March 21st, 2021) unless there are further comments between now and then.

**Comment by rlucas7** [comment]
Squashed + merged thanks @WarrenWeckesser + @mdhaber + @person142 I'll add a note to release docs/wiki

**Pull request merged**
